Add documentation for direct reach and frequency and volumetric measurement

iverson52000 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)
<!-- Reviewable:end -->

SanjayVas reviewed (approved):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @duliomatos1, @matthewclegg, @stevenwarejones, and @tristanvuong2021)



<!-- Sent from Reviewable.io -->

matthewclegg commented on src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto:
```diff
@@ -73,5 +73,6 @@ message MeasurementSpec {
     ReachAndFrequency reach_and_frequency = 4;
     Impression impression = 5;
     Duration duration = 6;
+    ReachAndFrequency direct_reach_and_frequency = 7;
```
One question I wonder about is whether it is a better design to leave this API as it is and to have the Kingdom decide whether to use the MPC or to send a direct measurement request to the EDP.  What are your thoughts on the pros and cons of this?

matthewclegg reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1, @iverson52000, and @stevenwarejones)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N4cJrcRCl7xmy29Pkoy:b1p9hk9) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

My only concern with that is that the MC and EDP lose out on their ability to consent to that behavior. From a technical perspective, the EDP needs to fulfill the requisition differently so that needs to be clear in the Requisition. Our consent signaling mechanism is such that all relevant parameters in the requisition must be signed by the MC.
</blockquote></details>

Sorry for not initially posting my comment in Reviewable.  That was careless.  I wonder if it would make sense for us to have a quick chat to understand this issue?  I would really like to hear Albert's thoughts on this too.


<!-- Sent from Reviewable.io -->

stevenwarejones reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N4gnNDO4LnGVijTK4hQ:b-teqxcl) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, iverson52000 (Albert Hsu) wrote…</i></summary><blockquote>

Sure I can set up a meeting today. I will invite @tristanvuong2021 too since I am mainly working with him on this task
</blockquote></details>

I tend to agree with Sanjay. Not sure we want the kingdom deciding right? Shouldn't that logic live in the Measurement consumer.


<!-- Sent from Reviewable.io -->

iverson52000 commented:
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N4h_Few45Dth52FHU8r:b-apccwx) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

I tend to agree with Sanjay. Not sure we want the kingdom deciding right? Shouldn't that logic live in the Measurement consumer.
</blockquote></details>

Hi @stevenwarejones. Yesterday we discussed this internally and our current thinking is that it is a better design to not expose direct reach measurement as part of the Kingdom API. So this new measurement_type may not be needed. The main reason is this is an internal performance optimization, and there is no benefit to the MC to having a choice between the two methods.  If you like, we'd be happy to sync up in a dev connect to discuss further.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N4hgiku1FcZEL6BMA0B:boecbtl) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, iverson52000 (Albert Hsu) wrote…</i></summary><blockquote>

Hi @stevenwarejones. Yesterday we discussed this internally and our current thinking is that it is a better design to not expose direct reach measurement as part of the Kingdom API. So this new measurement_type may not be needed. The main reason is this is an internal performance optimization, and there is no benefit to the MC to having a choice between the two methods.  If you like, we'd be happy to sync up in a dev connect to discuss further.
</blockquote></details>

To fill in what was discussed offline:

As @iverson52000 stated, direct measurements are really an optimization based on the observation that the MPC can be skipped in cases where a Measurement only includes a single EDP. In this case, an EDP can compute the Measurement on their own and fulfill the Requisition directly at the Kingdom, similar to how volumetric measurements work.

The alternative proposal was to have the fact that it's a direct measurement be noted in the Requisition outside of the signed MeasurementSpec. My main concern was whether the Kingdom could use this to learn things it shouldn't (the main reason behind the consent signaling paradigm). @matthewclegg reminded me that even for direct requisition fulfillment, the EDP still ends up encrypting things using the MC's key. Therefore, the Kingdom can't use this mechanism to learn anything new.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N4hiPUy8tBOjkd0ZMR4:bt6nebz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

To fill in what was discussed offline:

As @iverson52000 stated, direct measurements are really an optimization based on the observation that the MPC can be skipped in cases where a Measurement only includes a single EDP. In this case, an EDP can compute the Measurement on their own and fulfill the Requisition directly at the Kingdom, similar to how volumetric measurements work.

The alternative proposal was to have the fact that it's a direct measurement be noted in the Requisition outside of the signed MeasurementSpec. My main concern was whether the Kingdom could use this to learn things it shouldn't (the main reason behind the consent signaling paradigm). @matthewclegg reminded me that even for direct requisition fulfillment, the EDP still ends up encrypting things using the MC's key. Therefore, the Kingdom can't use this mechanism to learn anything new.
</blockquote></details>

Sorry, I forgot to mention the one "new" thing that an EDP would learn is that they are the only EDP in a given measurement. This is true regardless of whether the direct measurement bit is signed by the MC.


<!-- Sent from Reviewable.io -->

stevenwarejones reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N5LLeUf5CuCE1XORhO2:b-x63thj) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Sorry, I forgot to mention the one "new" thing that an EDP would learn is that they are the only EDP in a given measurement. This is true regardless of whether the direct measurement bit is signed by the MC.
</blockquote></details>

so, the edp would still fulfill the requisition to an mpc worker? We'd then need to have the MPC worker see that its a direct measurement and act accordingly?


<!-- Sent from Reviewable.io -->

stevenwarejones reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N6PG3OrBZR4_EFBl2fX:bsqvhrl) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

No. Direct reach would follow the same path as volumetric measurements, where the EDP instead sends the result directly to the Kingdom encrypted with the MC's public key. Other measurements, in contrast, are sent to a Duchy encrypted with the combined ElGamal public key of all of the Duchies.

The other difference with direct reach is that EDPs send an actual value rather than a sketch.
</blockquote></details>

what about calling this `single_publisher_reach_and_frequency`?


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N6nZ_ec9UxsHQUnqGNR:b-93i76k) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

what about calling this `single_publisher_reach_and_frequency`?
</blockquote></details>

Not a bad idea, though with `data_provider` instead of `publisher` as that's the term we use in the API.


<!-- Sent from Reviewable.io -->

matthewclegg reviewed (approved):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @tristanvuong2021)



<!-- Sent from Reviewable.io -->

duliomatos1 reviewed (approved):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @tristanvuong2021)



<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N6yH3K89AbQF98Q9nXz:b-qwl6o) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

@iverson52000, can you update this PR to reflect what was discussed above? In particular, not having this be a separate measurement type but instead signaling somewhere in the Requisition that the EDP needs to handle this differently?

I actually changed my mind on using "single-publisher" to distinguish this, as "direct" is already in the service method  we use for fulfilling requisitions for volumetric metrics (`FulfillDirectRequisition`). I assume these measurements would use the same method.
</blockquote></details>

From world-federation-of-advertisers/cross-media-measurement#619, it looks like we're signaling this to the EDP by the lack of a protocol config instead of having a separate field in the Requisition? Make sure that's clearly documented in `requisition.proto` then.

I don't think there's anywhere in the current API definition that indicates when the EDP should call `Requisitions.FulfillDirectRequisition` instead of `RequisitionFulfillment.FulfillRequisition`.


<!-- Sent from Reviewable.io -->

iverson52000 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N6yHeZn793AwgGu38mR:b-ic3eo5) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

From world-federation-of-advertisers/cross-media-measurement#619, it looks like we're signaling this to the EDP by the lack of a protocol config instead of having a separate field in the Requisition? Make sure that's clearly documented in `requisition.proto` then.

I don't think there's anywhere in the current API definition that indicates when the EDP should call `Requisitions.FulfillDirectRequisition` instead of `RequisitionFulfillment.FulfillRequisition`.
</blockquote></details>

@sanjayvas I was not planning to merge this PR since a new measurement_type is not needed anymore. But sounds I will need to add some documentation in ```requisition.proto```?
The actual change for direct measurement is in this PR: https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/619

The specific change is this line: https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/619/files#diff-5041a4afec93920965b3804e92a6651edea0416cd54c933ff94bf2dbeaca48c0R278


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N6zC12I4qUlA3sEVLpH:b-fmiu9i) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, iverson52000 (Albert Hsu) wrote…</i></summary><blockquote>

@sanjayvas I was not planning to merge this PR since a new measurement_type is not needed anymore. But sounds I will need to add some documentation in ```requisition.proto```?
The actual change for direct measurement is in this PR: https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/619

The specific change is this line: https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/619/files#diff-5041a4afec93920965b3804e92a6651edea0416cd54c933ff94bf2dbeaca48c0R278
</blockquote></details>

Yes, add some documentation (which might technically cover what was already implemented for volumetric measurements). You can drop this PR and do it in a different one if this one has too much unrelated history.


<!-- Sent from Reviewable.io -->

iverson52000 commented:
*[`src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto` line 76 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N4Y1oSIoFNh4MN3F7EH-r1-76:-N7MyyOr04huK93nzUmX:bem8mbr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/50e214921f55150ca52825aceed853115514743f/src/main/proto/wfa/measurement/api/v2alpha/measurement_spec.proto#L76)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Yes, add some documentation (which might technically cover what was already implemented for volumetric measurements). You can drop this PR and do it in a different one if this one has too much unrelated history.
</blockquote></details>

Updated the title of the PR and added additional documentation for direct and volumetric measurement in `requisition.proto`


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 2 unresolved discussions (waiting on @iverson52000 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/requisition.proto` line 59 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N7RXd_XDF93EbTx1yaZ:-N7RXd_XDF93EbTx1ya_:b-i7ybok) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/2e760c9a620a80bd6e1c7127710add3d58ef4042/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L59)):*
> ```protobuf
>   // Denormalized `protocol_config` field from `measurement`. Immutable.
>   //
>   // May not be set when `measurement_type` is `REACH_AND_FREQUENCY` and it only
> ```

This isn't the important part to document. What's important is that if this is set, the requisition can be fulfilled by calling `RequisitionFulfillment/FulfillRequisition`. Otherwise the`Requisition` is considered "direct" and can be fulfilled by calling `Requisitions/FulfillDirectRequisition`.



_Code quote:_
```protobuf
  // May not be set when `measurement_type` is `REACH_AND_FREQUENCY` and it only
  // has single EDP(1 DataProvider which is direct reach and frequency
  // measurement).
  // May not be set for volumetric measurement(`measurement_type` is
  // `IMPRESSION` or `DURATION`).
```


<!-- Sent from Reviewable.io -->

iverson52000 commented:
*[`src/main/proto/wfa/measurement/api/v2alpha/requisition.proto` line 59 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N7RXd_XDF93EbTx1yaZ:-N7RfGIKAngtL73OTSTK:b-3wvwwi) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/2e760c9a620a80bd6e1c7127710add3d58ef4042/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L59)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

This isn't the important part to document. What's important is that if this is set, the requisition can be fulfilled by calling `RequisitionFulfillment/FulfillRequisition`. Otherwise the`Requisition` is considered "direct" and can be fulfilled by calling `Requisitions/FulfillDirectRequisition`.
</blockquote></details>

Update based on the comment


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (approved):
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 2 unresolved discussions (waiting on @iverson52000 and @tristanvuong2021)

___
*[`src/main/proto/wfa/measurement/api/v2alpha/requisition.proto` line 61 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109#-N7Ro5MW9aRcJNIHuTYd:-N7Ro5MW9aRcJNIHuTYe:b-zay0rv) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/3caaa3bc7b012beb711ebd83541895be8b5552b6/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L61)):*
> ```protobuf
>   // If this is set, the requisition can be fulfilled by calling
>   // RequisitionFulfillment/FulfillRequisition(`measurement_type` is
>   // `REACH_AND_FREQUENCY` and it has multiple EDPs)
> ```

nit: the abbreviation "EDP" isn't used in the CMMS API. Instead we refer to the `DataProvider` resource type.



_Code quote:_
```protobuf
EDPs
```


<!-- Sent from Reviewable.io -->

iverson52000 reviewed (commented):
> Reviewed 1 of 2 files at r2, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: all files reviewed, 1 unresolved discussion (waiting on @tristanvuong2021)



<!-- Sent from Reviewable.io -->

matthewclegg reviewed (approved):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/109)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @iverson52000)



<!-- Sent from Reviewable.io -->
